File: go/quantum_resistant_kdf_mac.go
// Quantum Resistant - KDF and MAC
package pqc

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/sha512"

	"golang.org/x/crypto/argon2"
	"golang.org/x/crypto/bcrypt"
	"golang.org/x/crypto/hkdf"
	"golang.org/x/crypto/pbkdf2"
	"golang.org/x/crypto/scrypt"
	"golang.org/x/crypto/poly1305"
)

// KDF Functions
func HKDFDerive() {
	secret := []byte("secret key material")
	salt := []byte("random salt")
	info := []byte("context info")

	hkdfReader := hkdf.New(sha256.New, secret, salt, info)
	key := make([]byte, 32)
	hkdfReader.Read(key)
}

func PBKDF2Derive() {
	password := []byte("user password")
	salt := []byte("random salt")
	key := pbkdf2.Key(password, salt, 100000, 32, sha256.New)
	_ = key
}

func Argon2Derive() {
	password := []byte("user password")
	salt := []byte("random salt value")

	// Argon2id
	key := argon2.IDKey(password, salt, 1, 64*1024, 4, 32)
	_ = key
}

func ScryptDerive() {
	password := []byte("user password")
	salt := []byte("random salt")
	key, _ := scrypt.Key(password, salt, 32768, 8, 1, 32)
	_ = key
}

func BcryptHash() {
	password := []byte("user password")
	hash, _ := bcrypt.GenerateFromPassword(password, bcrypt.DefaultCost)
	err := bcrypt.CompareHashAndPassword(hash, password)
	_ = err
}

// MAC Functions
func HMACSHA256Mac() {
	key := []byte("secret key")
	message := []byte("message to authenticate")

	mac := hmac.New(sha256.New, key)
	mac.Write(message)
	tag := mac.Sum(nil)
	_ = tag
}

func HMACSHA512Mac() {
	key := []byte("secret key")
	message := []byte("message to authenticate")

	mac := hmac.New(sha512.New, key)
	mac.Write(message)
	tag := mac.Sum(nil)
	_ = tag
}

func Poly1305Mac() {
	var key [32]byte
	var out [16]byte
	message := []byte("message to authenticate")

	poly1305.Sum(&out, message, &key)
}

File: rust/quantum_resistant_kdf_mac.rs
// Quantum Resistant - KDF and MAC
use hkdf::Hkdf;
use sha2::Sha256;
use pbkdf2::pbkdf2_hmac;
use argon2::{Argon2, PasswordHasher};
use scrypt::{scrypt, Params as ScryptParams};
use hmac::{Hmac, Mac};

type HmacSha256 = Hmac<Sha256>;
type HmacSha512 = Hmac<sha2::Sha512>;

// KDF Functions
fn hkdf_derive() {
    let hk = Hkdf::<Sha256>::new(Some(&salt), &ikm);
    let mut okm = [0u8; 32];
    hk.expand(&info, &mut okm).unwrap();
}

fn pbkdf2_derive() {
    let mut key = [0u8; 32];
    pbkdf2_hmac::<Sha256>(password, salt, 100_000, &mut key);
}

fn argon2_derive() {
    let argon2 = Argon2::default();
    let password_hash = argon2.hash_password(password, &salt).unwrap();
}

fn scrypt_derive() {
    let params = ScryptParams::new(15, 8, 1).unwrap();
    let mut output = [0u8; 32];
    scrypt(password, salt, &params, &mut output).unwrap();
}

fn bcrypt_hash() {
    use bcrypt::{hash, verify, DEFAULT_COST};
    let hashed = hash(password, DEFAULT_COST).unwrap();
    let valid = verify(password, &hashed).unwrap();
}

// MAC Functions
fn hmac_sha256_mac() {
    let mut mac = HmacSha256::new_from_slice(key).unwrap();
    mac.update(message);
    let result = mac.finalize();
}

fn hmac_sha512_mac() {
    let mut mac = HmacSha512::new_from_slice(key).unwrap();
    mac.update(message);
    let tag = mac.finalize().into_bytes();
}

fn poly1305_mac() {
    use poly1305::Poly1305;
    let mac = Poly1305::new(key.into());
    let tag = mac.compute_unpadded(message);
}
